pkg/agent/openflow/pipeline.go

```diff
@@ -454,411 +456,600 @@ func (c *client) DeleteOFEntries(ofEntries []binding.OFEntry) error {
-		Cookie(c.cookieAllocator.Request(category).Raw()).
-		Done())
-	return
-}
-
-// connectionTrackFlows generates flows that redirect traffic to ct_zone and handle traffic according to ct_state:
-// 1) commit new connections to ct_zone(0xfff0) in the ConntrackCommitTable.
-// 2) Add ct_mark on the packet if it is sent to the switch from the host gateway.
-// 3) Allow traffic if it hits ct_mark and is sent from the host gateway.
-// 4) Drop all invalid traffic.
-// 5) Let other traffic go to the SessionAffinityTable first and then the ServiceLBTable.
-//    The SessionAffinityTable is a side-effect table which means traffic will not
-//    be resubmitted to any table. serviceLB does Endpoint selection for traffic
-//    to a Service.
-// 6) Add a flow to bypass reject response packet sent by the controller.
-func (c *client) connectionTrackFlows(category cookie.Category) []binding.Flow {
-	flows := c.conntrackBasicFlows(category)
-	if c.enableProxy {
-		// Replace the default flow with multiple resubmits actions.
-		if c.proxyAll {
-			flows = append(flows, ConntrackStateTable.BuildFlow(priorityMiss).
-				Cookie(c.cookieAllocator.Request(category).Raw()).
-				Action().ResubmitToTable(ServiceClassifierTable.GetID()).
-				Action().ResubmitToTable(SessionAffinityTable.GetID()).
-				Action().ResubmitToTable(ServiceLBTable.GetID()).
-				Done())
-		} else {
-			flows = append(flows, ConntrackStateTable.BuildFlow(priorityMiss).
-				Cookie(c.cookieAllocator.Request(category).Raw()).
-				Action().ResubmitToTable(SessionAffinityTable.GetID()).
-				Action().ResubmitToTable(ServiceLBTable.GetID()).
-				Done())
-		}
-
-		for _, proto := range c.ipProtocols {
-			gatewayIP := c.nodeConfig.GatewayConfig.IPv4
-			serviceVirtualIP := config.VirtualServiceIPv4
-			snatZone := SNATCtZone
-			ctZone := CtZone
-			if proto == binding.ProtocolIPv6 {
-				gatewayIP = c.nodeConfig.GatewayConfig.IPv6
-				serviceVirtualIP = config.VirtualServiceIPv6
-				snatZone = SNATCtZoneV6
-				ctZone = CtZoneV6
-			}
+// Feature: PodConnectivity
+// Stage: ClassifierStage
+// Tables: ClassifierTable
+// Refactored from:
+//   - func (c *client) podUplinkClassifierFlows(dstMAC net.HardwareAddr, category cookie.Category) (flows []binding.Flow)
+// podUplinkClassifierFlows generates the flows to mark the packets with target destination MAC address from uplink/bridge
+// port, which are needed when uplink is connected to OVS bridge and Antrea IPAM is configured.
+func (f *featurePodConnectivity) podUplinkClassifierFlows(dstMAC net.HardwareAddr, vlanID *uint16) []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var flows []binding.Flow
+	ctZoneTypeRegMark := IPCtZoneTypeRegMark
+	if len(f.ipProtocols) == 1 && f.ipProtocols[0] == binding.ProtocolIPv6 {
+		ctZoneTypeRegMark = IPv6CtZoneTypeRegMark
+	}
+	nonVLAN := true
+	vlanIDValue := uint16(0)
+	if vlanID != nil && *vlanID > 0 {
+		nonVLAN = false
+		vlanIDValue = *vlanID
+	}
+	flows = append(flows,
```

Comment:
@jianjuns If we want to use MatchNonVLAN, we must add `if...else...` for below 4 flows. Thus it makes the code untidy.

Reply:
Ok. Then we can go your proposal of using VLAN 0 to indicate no VLAN?

Reply:
Yes, VLAN 0 means no VLAN at most of functions. However, for `MatchVLAN` function, I think we should still keep the `nonVLAN` parameter because `MatchVLAN` supports mask.
Below call means match packets with any VLAN.
vlanMask := 0x1000
MatchVLAN(false, 0, &vlanMask)

Reply:
I guess we do not really have a case to match vlanMask with VLAN 0? Even we really have that, we can check: vlan == 0 && vlanMask == nil to know it is nonVLAN?

BTW, could we use an integer for vlanMask too instead of a pointer? Could we check vlanMask == 0 for noMask?

Reply:
Yes, we can use this method, but I think this is not intuitive enough.
@wenyingd Do you have any opinion for below implementation of `MatchVLAN` function?
`MatchVLAN(vlanID uint16, vlanMask uint16)` vlanMask=0 means no mask, and vlanID=0&&vlanMask=0 means match nonVLAN.
We'll support below cases in `MatchVLAN`, except `vlan=0 no mask` which Antrea won't support.
```
              Value 0x0000 (OFPVID_NONE), mask 0xffff (or no mask)
                     Matches only packets without an 802.1Q header.

              Value 0x1000, mask 0x1000
                     Matches  any  packet with an 802.1Q header, regardless of
                     VLAN ID.

              Value 0x1009, mask 0xffff (or no mask)
                     Match only packets with an 802.1Q header with VLAN ID 9.

              Value 0x1001, mask 0x1001
                     Matches only packets that have an 802.1Q header  with  an
                     odd-numbered  VLAN  ID. (This is just an example; one can
                     match on any desired VLAN ID bit pattern.)
```